tpcc report "failed Error 8133: data inconsistency in table: customer, index: PRIMARY, index-count:1 != record-count:0\r\n" during tikv balance

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
./tikv-server -V
 TiKV 
Release Version:   8.2.0-alpha
Edition:           Community
Git Commit Hash:   212d2d51de0abb9382057aca71f7dde257044c8d
Git Commit Branch: heads/refs/tags/v8.2.0-alpha
UTC Build Time:    2024-07-05 11:37:37
Rust Version:      rustc 1.77.0-nightly (89e2160c4 2023-12-27)
Enable Features:   memory-engine pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine trace-async-tasks openssl-vendored
Profile:           dist_release

### What operating system and CPU are you using?
8c/32g

### Steps to reproduce
1、run tpcc
2、config set max-store-down-time 10m
2、down one of tikv last for 20mins

### What did you expect?
no unexpected error

### What did happened?
tpcc report error during tikv balance after fault recover
[2024/07/07 07:45:51.990 +08:00] [INFO] [workload.go:680] ["[2024-07-06 23:45:51] execute run failed, err exec SELECT c_discount, c_last, c_credit, w_tax FROM customer, warehouse WHERE w_id = ? AND c_w_id = w_id AND c_d_id = ? AND c_id = ?(wID=863,dID=5,cID=2731) failed Error 8133: data inconsistency in table: customer, index: PRIMARY, index-count:1 != record-count:0\r\n"]

![d655b8a0-c443-402e-bcd8-5893afe693bb](https://github.com/tikv/tikv/assets/84712107/a0982dca-caac-483f-9f47-3591db05ab46)

tidb error logs:

```log
[2024/07/12 05:03:22.729 +08:00] [ERROR] [reporter.go:213] ["indexLookup found data inconsistency"] [conn=3573547388] [session_alias=] [table_name=customer] [index_name=PRIMARY] [index_cnt=1] [table_cnt=0] [missing_handles="[17087983]"] [total_handles="[17087983]"] ...
```

After check,  found a lot of rows data of table `customer` are missing:

```sql
tpcc> select count(1) from customer use index(primary);
+----------+
| count(1) |
+----------+
| 30000000 |
+----------+
tpcc> select count(1) from customer use index(idx_customer);
+----------+
| count(1) |
+----------+
| 30000000 |
+----------+
tpcc> select count(1) from customer ignore index(primary, idx_customer);
+----------+
| count(1) |
+----------+
| 29986907 |
+----------+

-- query the missing row's _tidb_rowid
tpcc> select _tidb_rowid, c_w_id, c_d_id, c_id from customer use index(primary) where _tidb_rowid not in (select _tidb_rowid from customer ignore index(primary, idx_customer)) order by _tidb_rowid;
+-------------+--------+--------+------+
| _tidb_rowid | c_w_id | c_d_id | c_id |
+-------------+--------+--------+------+
| 17082740    | 570    | 5      | 740  |
| 17082741    | 570    | 5      | 741  |
| 17082742    | 570    | 5      | 742  |

...
...
...
| 17096438    | 570    | 9      | 2438 |
| 17096439    | 570    | 9      | 2439 |
| 17096440    | 570    | 9      | 2440 |
+-------------+--------+--------+------+
13093 rows in set
```

Based on the missing row data _tidb_rowid range [17082740, 17096440], locate Region 31434961, which is on Store 5, which is tikv-2:

```sql
>  show table customer regions;
...
 31434959  | t_111_r_17076872 | t_111_r_17108437 | 31434961  | 5 | 31434960, 31434961, 31434962 | 0
```

Looking for the tikv-2 logs around the time `2024/07/12 05:03:22`, we found that there is an apply snapshot and 2 delete data logs, and the `tid: 111` behind the log is the table id 111 parsed according to the key, that is, the table id of the `customer` table, start_handle and end_handle are also decode from the key. The following three log messages are:

- 05:02:33.138,apply snapshot, the corresponding data is the row data of the customer table: [17082740,17096441) 
- 05:02:57.400, delete data, the corresponding data is the row data of the customer table: [17082740, 17087300)
- 05:02:57.400, delete data, The corresponding data is the row data of the customer table: [17087897, 17096441) 

Note that the missing row data _tidb_rowid range of the table `customer` is [17082740, 17096440], which is exactly the data range of the deleted data above.

```log
[2024/07/12 05:02:33.138 +08:00] [INFO] [peer_storage.rs:731] ["apply snapshot with state ok"] [for_witness=false] [state="applied_index: 142 truncated_state { index: 142 term: 7 }"] [region="id: 31416981 start_key: 7480000000000000FF6F5F728000000001FF04A9740000000000FA end_key: 7480000000000000FF6F5F728000000001FF04DEF90000000000FA region_epoch { conf_ver: 96 version: 1223 } peers { id: 31416982 store_id: 1 } peers { id: 31416984 store_id: 4 } peers { id: 31425373 store_id: 6 } peers { id: 31426272 store_id: 5 role: Learner }"] [peer_id=31426272] [region_id=31416981] [thread_id=103] tid: 111, start_handle: 17082740, end_handle: 17096441, contain_suspect: true
[2024/07/12 05:02:57.400 +08:00] [INFO] [region.rs:472] ["delete data in range because of stale"] [end_key=7A7480000000000000FF6F5F728000000001FF04BB440000000000FA] [start_key=7A7480000000000000FF6F5F728000000001FF04A9740000000000FA] [region_id=31405673] [thread_id=99] tid: 111, start_handle: 17082740, end_handle: 17087300, contain_suspect: true
[2024/07/12 05:02:57.400 +08:00] [INFO] [region.rs:472] ["delete data in range because of stale"] [end_key=7A7480000000000000FF6F5F728000000001FF04DEF90000000000FA] [start_key=7A7480000000000000FF6F5F728000000001FF04BD990000000000FA] [region_id=31406521] [thread_id=99] tid: 111, start_handle: 17087897, end_handle: 17096441, contain_suspect: true
```

--- a/components/raftstore/src/store/worker/region.rs
+++ b/components/raftstore/src/store/worker/region.rs
@@ -10,7 +10,7 @@ use std::{
     sync::{
         atomic::{AtomicBool, AtomicUsize, Ordering},
         mpsc::SyncSender,
-        Arc, Mutex,
+        Arc,
     },
     time::Duration,
     u64,
@@ -341,12 +341,22 @@ where
     }
 }
 
-struct RegionCleaner<EK>
+pub struct Runner<EK, R, T>
 where
     EK: KvEngine,
+    T: PdClient + 'static,
 {
+    batch_size: usize,
     use_delete_range: bool,
-    engine: EK,
+    ingest_copy_symlink: bool,
+    clean_stale_tick: usize,
+    clean_stale_check_interval: Duration,
+    clean_stale_ranges_tick: usize,
+
+    tiflash_stores: HashMap<u64, bool>,
+    // we may delay some apply tasks if level 0 files to write stall threshold,
+    // pending_applies records all delayed apply task, and will check again later
+    pending_applies: VecDeque<Task<EK::Snapshot>>,
     // Ranges that have been logically destroyed at a specific sequence number. We can
     // assume there will be no reader (engine snapshot) newer than that sequence number. Therefore,
     // they can be physically deleted with `DeleteFiles` when we're sure there is no older
@@ -356,13 +366,206 @@ where
     // The sole purpose of maintaining this list is to optimize deletion with `DeleteFiles`
     // whenever we can. Errors while processing them can be ignored.
     pending_delete_ranges: PendingDeleteRanges,
+
+    engine: EK,
     mgr: SnapManager,
+    coprocessor_host: CoprocessorHost<EK>,
+    router: R,
+    pd_client: Option<Arc<T>>,
+    pool: FuturePool,
 }
 
-impl<EK> RegionCleaner<EK>
+impl<EK, R, T> Runner<EK, R, T>
 where
     EK: KvEngine,
+    R: CasualRouter<EK>,
+    T: PdClient + 'static,
 {
+    pub fn new(
+        engine: EK,
+        mgr: SnapManager,
+        cfg: Arc<VersionTrack<Config>>,
+        coprocessor_host: CoprocessorHost<EK>,
+        router: R,
+        pd_client: Option<Arc<T>>,
+    ) -> Runner<EK, R, T> {
+        Runner {
+            batch_size: cfg.value().snap_apply_batch_size.0 as usize,
+            use_delete_range: cfg.value().use_delete_range,
+            ingest_copy_symlink: cfg.value().snap_apply_copy_symlink,
+            clean_stale_tick: 0,
+            clean_stale_check_interval: Duration::from_millis(
+                cfg.value().region_worker_tick_interval.as_millis(),
+            ),
+            clean_stale_ranges_tick: cfg.value().clean_stale_ranges_tick,
+            tiflash_stores: HashMap::default(),
+            pending_applies: VecDeque::new(),
+            pending_delete_ranges: PendingDeleteRanges::default(),
+            engine,
+            mgr,
+            coprocessor_host,
+            router,
+            pd_client,
+            pool: YatpPoolBuilder::new(DefaultTicker::default())
+                .name_prefix("snap-generator")
+                .thread_count(
+                    1,
+                    cfg.value().snap_generator_pool_size,
+                    SNAP_GENERATOR_MAX_POOL_SIZE,
+                )
+                .build_future_pool(),
+        }
+    }
+
+    pub fn snap_generator_pool(&self) -> FuturePool {
+        self.pool.clone()
+    }
+
+    fn region_state(&self, region_id: u64) -> Result<RegionLocalState> {
+        let region_key = keys::region_state_key(region_id);
+        let region_state: RegionLocalState =
+            match box_try!(self.engine.get_msg_cf(CF_RAFT, &region_key)) {
+                Some(state) => state,
+                None => {
+                    return Err(box_err!(
+                        "failed to get region_state from {}",
+                        log_wrappers::Value::key(&region_key)
+                    ));
+                }
+            };
+        Ok(region_state)
+    }
+
+    fn apply_state(&self, region_id: u64) -> Result<RaftApplyState> {
+        let state_key = keys::apply_state_key(region_id);
+        let apply_state: RaftApplyState =
+            match box_try!(self.engine.get_msg_cf(CF_RAFT, &state_key)) {
+                Some(state) => state,
+                None => {
+                    return Err(box_err!(
+                        "failed to get apply_state from {}",
+                        log_wrappers::Value::key(&state_key)
+                    ));
+                }
+            };
+        Ok(apply_state)
+    }
+
+    /// Applies snapshot data of the Region.
+    fn apply_snap(&mut self, region_id: u64, peer_id: u64, abort: Arc<AtomicUsize>) -> Result<()> {
+        info!("begin apply snap data"; "region_id" => region_id, "peer_id" => peer_id);
+        fail_point!("region_apply_snap", |_| { Ok(()) });
+        fail_point!("region_apply_snap_io_err", |_| {
+            Err(crate::store::SnapError::Other(box_err!("io error")))
+        });
+        check_abort(&abort)?;
+
+        let mut region_state = self.region_state(region_id)?;
+        let region = region_state.get_region().clone();
+
+        let start_key = keys::enc_start_key(&region);
+        let end_key = keys::enc_end_key(&region);
+        check_abort(&abort)?;
+        self.clean_overlap_ranges(start_key, end_key)?;
+        check_abort(&abort)?;
+        fail_point!("apply_snap_cleanup_range");
+
+        // apply snapshot
+        let apply_state = self.apply_state(region_id)?;
+        let term = apply_state.get_truncated_state().get_term();
+        let idx = apply_state.get_truncated_state().get_index();
+        let snap_key = SnapKey::new(region_id, term, idx);
+        self.mgr.register(snap_key.clone(), SnapEntry::Applying);
+        defer!({
+            self.mgr.deregister(&snap_key, &SnapEntry::Applying);
+        });
+        let mut s = box_try!(self.mgr.get_snapshot_for_applying(&snap_key));
+        if !s.exists() {
+            return Err(box_err!("missing snapshot file {}", s.path()));
+        }
+        check_abort(&abort)?;
+        let timer = Instant::now();
+        let options = ApplyOptions {
+            db: self.engine.clone(),
+            region: region.clone(),
+            abort: Arc::clone(&abort),
+            write_batch_size: self.batch_size,
+            coprocessor_host: self.coprocessor_host.clone(),
+            ingest_copy_symlink: self.ingest_copy_symlink,
+        };
+        s.apply(options)?;
+        self.coprocessor_host
+            .post_apply_snapshot(&region, peer_id, &snap_key, Some(&s));
+
+        // delete snapshot state.
+        let mut wb = self.engine.write_batch();
+        region_state.set_state(PeerState::Normal);
+        box_try!(wb.put_msg_cf(CF_RAFT, &keys::region_state_key(region_id), &region_state));
+        box_try!(wb.delete_cf(CF_RAFT, &keys::snapshot_raft_state_key(region_id)));
+        wb.write().unwrap_or_else(|e| {
+            panic!("{} failed to save apply_snap result: {:?}", region_id, e);
+        });
+        info!(
+            "apply new data";
+            "region_id" => region_id,
+            "time_takes" => ?timer.saturating_elapsed(),
+        );
+        Ok(())
+    }
+
+    /// Tries to apply the snapshot of the specified Region. It calls
+    /// `apply_snap` to do the actual work.
+    fn handle_apply(&mut self, region_id: u64, peer_id: u64, status: Arc<AtomicUsize>) {
+        let _ = status.compare_exchange(
+            JOB_STATUS_PENDING,
+            JOB_STATUS_RUNNING,
+            Ordering::SeqCst,
+            Ordering::SeqCst,
+        );
+        SNAP_COUNTER.apply.start.inc();
+
+        let start = Instant::now();
+
+        let tombstone = match self.apply_snap(region_id, peer_id, Arc::clone(&status)) {
+            Ok(()) => {
+                status.swap(JOB_STATUS_FINISHED, Ordering::SeqCst);
+                SNAP_COUNTER.apply.success.inc();
+                false
+            }
+            Err(Error::Abort) => {
+                warn!("applying snapshot is aborted"; "region_id" => region_id);
+                self.coprocessor_host
+                    .cancel_apply_snapshot(region_id, peer_id);
+                assert_eq!(
+                    status.swap(JOB_STATUS_CANCELLED, Ordering::SeqCst),
+                    JOB_STATUS_CANCELLING
+                );
+                SNAP_COUNTER.apply.abort.inc();
+                // The snapshot is applied abort, it's not necessary to tombstone the peer.
+                false
+            }
+            Err(e) => {
+                warn!("failed to apply snap!!!"; "region_id" => region_id, "err" => %e);
+                self.coprocessor_host
+                    .cancel_apply_snapshot(region_id, peer_id);
+                status.swap(JOB_STATUS_FAILED, Ordering::SeqCst);
+                SNAP_COUNTER.apply.fail.inc();
+                // As the snapshot failed, the related peer should be marked tombstone.
+                // And as for the abnormal snapshot, it will be automatically cleaned up by
+                // the CleanupWorker later.
+                true
+            }
+        };
+
+        SNAP_HISTOGRAM
+            .apply
+            .observe(start.saturating_elapsed_secs());
+        let _ = self.router.send(
+            region_id,
+            CasualMessage::SnapshotApplied { peer_id, tombstone },
+        );
+    }
+
     /// Tries to clean up files in pending ranges overlapping with the given
     /// bounds. These pending ranges will be removed. Returns an updated range
     /// that also includes these ranges. Caller must ensure the remaining keys
@@ -529,262 +732,30 @@ where
         let wopts = WriteOptions::default();
         for cf in self.engine.cf_names() {
             // CF_LOCK usually contains fewer keys than other CFs, so we delete them by key.
-            let strategy = if cf == CF_LOCK {
-                DeleteStrategy::DeleteByKey
-            } else if self.use_delete_range {
-                DeleteStrategy::DeleteByRange
-            } else {
-                DeleteStrategy::DeleteByWriter {
-                    sst_path: self.mgr.get_temp_path_for_ingest(),
-                }
-            };
-            box_try!(self.engine.delete_ranges_cf(&wopts, cf, strategy, ranges));
-        }
-        Ok(())
-    }
-}
-
-pub struct Runner<EK, R, T>
-where
-    EK: KvEngine,
-    T: PdClient + 'static,
-{
-    batch_size: usize,
-    ingest_copy_symlink: bool,
-    clean_stale_tick: usize,
-    clean_stale_check_interval: Duration,
-    clean_stale_ranges_tick: usize,
-
-    tiflash_stores: HashMap<u64, bool>,
-    // we may delay some apply tasks if level 0 files to write stall threshold,
-    // pending_applies records all delayed apply task, and will check again later
-    pending_applies: VecDeque<Task<EK::Snapshot>>,
-
-    engine: EK,
-    mgr: SnapManager,
-    coprocessor_host: CoprocessorHost<EK>,
-    router: R,
-    pd_client: Option<Arc<T>>,
-    snap_gen_pool: FuturePool,
-    region_cleanup_pool: FuturePool,
-    region_cleaner: Arc<Mutex<RegionCleaner<EK>>>,
-}
-
-impl<EK, R, T> Runner<EK, R, T>
-where
-    EK: KvEngine,
-    R: CasualRouter<EK>,
-    T: PdClient + 'static,
-{
-    pub fn new(
-        engine: EK,
-        mgr: SnapManager,
-        cfg: Arc<VersionTrack<Config>>,
-        coprocessor_host: CoprocessorHost<EK>,
-        router: R,
-        pd_client: Option<Arc<T>>,
-    ) -> Runner<EK, R, T> {
-        Runner {
-            batch_size: cfg.value().snap_apply_batch_size.0 as usize,
-            ingest_copy_symlink: cfg.value().snap_apply_copy_symlink,
-            clean_stale_tick: 0,
-            clean_stale_check_interval: Duration::from_millis(
-                cfg.value().region_worker_tick_interval.as_millis(),
-            ),
-            clean_stale_ranges_tick: cfg.value().clean_stale_ranges_tick,
-            tiflash_stores: HashMap::default(),
-            pending_applies: VecDeque::new(),
-            engine: engine.clone(),
-            mgr: mgr.clone(),
-            coprocessor_host,
-            router,
-            pd_client,
-            snap_gen_pool: YatpPoolBuilder::new(DefaultTicker::default())
-                .name_prefix("snap-generator")
-                .thread_count(
-                    1,
-                    cfg.value().snap_generator_pool_size,
-                    SNAP_GENERATOR_MAX_POOL_SIZE,
+            let (strategy, observer) = if cf == CF_LOCK {
+                (
+                    DeleteStrategy::DeleteByKey,
+                    &CLEAR_OVERLAP_REGION_DURATION.by_key,
                 )
-                .build_future_pool(),
-            region_cleanup_pool: YatpPoolBuilder::new(DefaultTicker::default())
-                .name_prefix("region-cleanup")
-                .thread_count(1, 1, 1)
-                .build_future_pool(),
-            region_cleaner: Arc::new(Mutex::new(RegionCleaner {
-                use_delete_range: cfg.value().use_delete_range,
-                engine,
-                pending_delete_ranges: PendingDeleteRanges::default(),
-                mgr,
-            })),
-        }
-    }
-
-    pub fn snap_generator_pool(&self) -> FuturePool {
-        self.snap_gen_pool.clone()
-    }
-
-    fn region_state(&self, region_id: u64) -> Result<RegionLocalState> {
-        let region_key = keys::region_state_key(region_id);
-        let region_state: RegionLocalState =
-            match box_try!(self.engine.get_msg_cf(CF_RAFT, &region_key)) {
-                Some(state) => state,
-                None => {
-                    return Err(box_err!(
-                        "failed to get region_state from {}",
-                        log_wrappers::Value::key(&region_key)
-                    ));
-                }
+            } else if self.use_delete_range {
+                (
+                    DeleteStrategy::DeleteByRange,
+                    &CLEAR_OVERLAP_REGION_DURATION.by_range,
+                )
+            } else {
+                (
+                    DeleteStrategy::DeleteByWriter {
+                        sst_path: self.mgr.get_temp_path_for_ingest(),
+                    },
+                    &CLEAR_OVERLAP_REGION_DURATION.by_ingest_files,
+                )
             };
-        Ok(region_state)
-    }
-
-    fn apply_state(&self, region_id: u64) -> Result<RaftApplyState> {
-        let state_key = keys::apply_state_key(region_id);
-        let apply_state: RaftApplyState =
-            match box_try!(self.engine.get_msg_cf(CF_RAFT, &state_key)) {
-                Some(state) => state,
-                None => {
-                    return Err(box_err!(
-                        "failed to get apply_state from {}",
-                        log_wrappers::Value::key(&state_key)
-                    ));
-                }
-            };
-        Ok(apply_state)
-    }
-
-    /// Applies snapshot data of the Region.
-    fn apply_snap(&mut self, region_id: u64, peer_id: u64, abort: Arc<AtomicUsize>) -> Result<()> {
-        info!("begin apply snap data"; "region_id" => region_id, "peer_id" => peer_id);
-        fail_point!("region_apply_snap", |_| { Ok(()) });
-        fail_point!("region_apply_snap_io_err", |_| {
-            Err(crate::store::SnapError::Other(box_err!("io error")))
-        });
-        check_abort(&abort)?;
-
-        let mut region_state = self.region_state(region_id)?;
-        let region = region_state.get_region().clone();
-
-        let start_key = keys::enc_start_key(&region);
-        let end_key = keys::enc_end_key(&region);
-        check_abort(&abort)?;
-        {
-            let mut region_cleaner = self.region_cleaner.lock().unwrap();
-            region_cleaner.clean_overlap_ranges(start_key, end_key)?;
+            let start = Instant::now();
+            box_try!(self.engine.delete_ranges_cf(&wopts, cf, strategy, ranges));
+            observer.observe(start.saturating_elapsed_secs());
         }
-        check_abort(&abort)?;
-        fail_point!("apply_snap_cleanup_range");
 
-        // apply snapshot
-        let apply_state = self.apply_state(region_id)?;
-        let term = apply_state.get_truncated_state().get_term();
-        let idx = apply_state.get_truncated_state().get_index();
-        let snap_key = SnapKey::new(region_id, term, idx);
-        self.mgr.register(snap_key.clone(), SnapEntry::Applying);
-        defer!({
-            self.mgr.deregister(&snap_key, &SnapEntry::Applying);
-        });
-        let mut s = box_try!(self.mgr.get_snapshot_for_applying(&snap_key));
-        if !s.exists() {
-            return Err(box_err!("missing snapshot file {}", s.path()));
-        }
-        check_abort(&abort)?;
-        let timer = Instant::now();
-        let options = ApplyOptions {
-            db: self.engine.clone(),
-            region: region.clone(),
-            abort: Arc::clone(&abort),
-            write_batch_size: self.batch_size,
-            coprocessor_host: self.coprocessor_host.clone(),
-            ingest_copy_symlink: self.ingest_copy_symlink,
-        };
-        s.apply(options)?;
-        self.coprocessor_host
-            .post_apply_snapshot(&region, peer_id, &snap_key, Some(&s));
-
-        // delete snapshot state.
-        let mut wb = self.engine.write_batch();
-        region_state.set_state(PeerState::Normal);
-        box_try!(wb.put_msg_cf(CF_RAFT, &keys::region_state_key(region_id), &region_state));
-        box_try!(wb.delete_cf(CF_RAFT, &keys::snapshot_raft_state_key(region_id)));
-        wb.write().unwrap_or_else(|e| {
-            panic!("{} failed to save apply_snap result: {:?}", region_id, e);
-        });
-        info!(
-            "apply new data";
-            "region_id" => region_id,
-            "time_takes" => ?timer.saturating_elapsed(),
-        );
         Ok(())
-    }
-
-    /// Tries to apply the snapshot of the specified Region. It calls
-    /// `apply_snap` to do the actual work.
-    fn handle_apply(&mut self, region_id: u64, peer_id: u64, status: Arc<AtomicUsize>) {
-        let _ = status.compare_exchange(
-            JOB_STATUS_PENDING,
-            JOB_STATUS_RUNNING,
-            Ordering::SeqCst,
-            Ordering::SeqCst,
-        );
-        SNAP_COUNTER.apply.start.inc();
-
-        let start = Instant::now();
-
-        let tombstone = match self.apply_snap(region_id, peer_id, Arc::clone(&status)) {
-            Ok(()) => {
-                status.swap(JOB_STATUS_FINISHED, Ordering::SeqCst);
-                SNAP_COUNTER.apply.success.inc();
-                false
-            }
-            Err(Error::Abort) => {
-                warn!("applying snapshot is aborted"; "region_id" => region_id);
-                self.coprocessor_host
-                    .cancel_apply_snapshot(region_id, peer_id);
-                assert_eq!(
-                    status.swap(JOB_STATUS_CANCELLED, Ordering::SeqCst),
-                    JOB_STATUS_CANCELLING
-                );
-                SNAP_COUNTER.apply.abort.inc();
-                // The snapshot is applied abort, it's not necessary to tombstone the peer.
-                false
-            }
-            Err(e) => {
-                warn!("failed to apply snap!!!"; "region_id" => region_id, "err" => %e);
-                self.coprocessor_host
-                    .cancel_apply_snapshot(region_id, peer_id);
-                status.swap(JOB_STATUS_FAILED, Ordering::SeqCst);
-                SNAP_COUNTER.apply.fail.inc();
-                // As the snapshot failed, the related peer should be marked tombstone.
-                // And as for the abnormal snapshot, it will be automatically cleaned up by
-                // the CleanupWorker later.
-                true
-            }
-        };
-
-        SNAP_HISTOGRAM
-            .apply
-            .observe(start.saturating_elapsed_secs());
-        let _ = self.router.send(
-            region_id,
-            CasualMessage::SnapshotApplied { peer_id, tombstone },
-        );
-    }
-
-    /// Checks the number of files at level 0 to avoid write stall after
-    /// ingesting sst. Returns true if the ingestion causes write stall.
-    fn ingest_maybe_stall(&self) -> bool {
-        for cf in SNAPSHOT_CFS {
-            // no need to check lock cf
-            if plain_file_used(cf) {
-                continue;
-            }
-            if self.engine.ingest_maybe_slowdown_writes(cf).expect("cf") {
-                return true;
-            }
-        }
-        false
     }
 
     /// Calls observer `pre_apply_snapshot` for every task.
@@ -920,7 +891,7 @@ where
                     start: UnixSecs::now(),
                 };
                 let scheduled_time = Instant::now_coarse();
-                self.snap_gen_pool.spawn(async move {
+                self.pool.spawn(async move {
                     SNAP_GEN_WAIT_DURATION_HISTOGRAM
                         .observe(scheduled_time.saturating_elapsed_secs());
 
@@ -961,19 +932,11 @@ where
                 start_key,
                 end_key,
             } => {
-                let region_cleaner = self.region_cleaner.clone();
-                self.region_cleanup_pool
-                    .spawn(async move {
-                        fail_point!("on_region_worker_destroy", region_id == 1000, |_| {});
-                        let mut region_cleaner = region_cleaner.lock().unwrap();
-                        // try to delay the range deletion because
-                        // there might be a coprocessor request related to this range
-                        region_cleaner.insert_pending_delete_range(region_id, start_key, end_key);
-                        region_cleaner.clean_stale_ranges();
-                    })
-                    .unwrap_or_else(|e| {
-                        error!("failed to destroy region"; "region_id" => region_id, "err" => ?e);
-                    });
+                fail_point!("on_region_worker_destroy", true, |_| {});
+                // try to delay the range deletion because
+                // there might be a coprocessor request related to this range
+                self.insert_pending_delete_range(region_id, start_key, end_key);
+                self.clean_stale_ranges();
             }
         }
     }
@@ -989,7 +952,7 @@ where
         self.handle_pending_applies(true);
         self.clean_stale_tick += 1;
         if self.clean_stale_tick >= self.clean_stale_ranges_tick {
-            self.region_cleaner.lock().unwrap().clean_stale_ranges();
+            self.clean_stale_ranges();
             self.clean_stale_tick = 0;
         }
     }

--- a/tests/failpoints/cases/test_merge.rs
+++ b/tests/failpoints/cases/test_merge.rs
@@ -1787,50 +1787,6 @@ fn test_destroy_source_peer_while_merging() {
     }
 }
 
-// Test that a store is able to generate snapshots while destroying a region. It
-// confirms the resolution of issue #12587.
-#[test]
-fn test_snap_gen_while_destroy_is_stuck() {
-    let mut cluster = new_node_cluster(0, 5);
-    configure_for_merge(&mut cluster.cfg);
-    let pd_client = Arc::clone(&cluster.pd_client);
-    pd_client.disable_default_operator();
-
-    cluster.run();
-
-    cluster.must_put(b"k1", b"v1");
-    cluster.must_put(b"k3", b"v3");
-    for i in 1..=5 {
-        must_get_equal(&cluster.get_engine(i), b"k1", b"v1");
-        must_get_equal(&cluster.get_engine(i), b"k3", b"v3");
-    }
-    cluster.must_split(&pd_client.get_region(b"k1").unwrap(), b"k2");
-
-    let region_1 = pd_client.get_region(b"k1").unwrap();
-    let region_2 = pd_client.get_region(b"k3").unwrap();
-    // Ensure that store 5 is the leader of region_1 and store 1 is the leader of
-    // region_2.
-    cluster.must_transfer_leader(region_1.get_id(), new_peer(5, 1001));
-    cluster.must_transfer_leader(region_2.get_id(), new_peer(1, 1));
-
-    let on_region_worker_destroy_fp = "on_region_worker_destroy";
-    fail::cfg(on_region_worker_destroy_fp, "pause").unwrap();
-
-    // Remove region_1 from from store 1 so that a destory task is scheduled.
-    // The destory task will be stuck due to the fail point above to simulate a
-    // slow destory scenario.
-    pd_client.must_remove_peer(region_1.get_id(), find_peer(&region_1, 1).unwrap().clone());
-    must_get_equal(&cluster.get_engine(1), b"k1", b"v1");
-
-    // Trigger a snapshot gen task on store 1 by removing and adding region_2 on
-    // store 5. The snapshot generation should go through without issues.
-    pd_client.must_remove_peer(region_2.get_id(), new_peer(5, 5));
-    must_get_none(&cluster.get_engine(5), b"k3");
-
-    pd_client.must_add_peer(region_2.get_id(), new_peer(5, 6));
-    must_get_equal(&cluster.get_engine(5), b"k3", b"v3");
-}
-
 struct MsgTimeoutFilter {
     // wrap with mutex to make tx Sync.
     tx: Mutex<Sender<RaftMessage>>,

